Implement compose pass

PoignardAzur opened the pull request:
Add `Widget::compose` method.
Add `ComposeCtx` type.
Add `needs_compose`, `request_compose`, `translation_changed` flags to WidgetState.
Add `window_origin` attribute to WidgetState.
Add convenience methods to TreeArena.
Remove `needs_window_origin` flag and `parent_window_origin` attribute.

waywardmonkeys commented on masonry/src/widget/widget_state.rs:
```diff
@@ -5,11 +5,17 @@
 
 use std::sync::atomic::{AtomicBool, Ordering};
 
+use kurbo::Vec2;
```
`use crate::kurbo::Vec2;`

waywardmonkeys replied:
Actually, can combine with the import below.

Philipp-M replied:
(I mean it's just a nit, but I agree, it's *slightly* cleaner, same for the other `use` comments here)

PoignardAzur replied:
Won't fix, see other comment.

Philipp-M replied:
We should probably generally find a unique style in the future, just to be consistent.

waywardmonkeys replied:
We've had an agreement in the past in #223, but I needed to rebase it and it slipped through the cracks.

I've resurrected it.

waywardmonkeys reviewed (commented):
Some initial comments...

PoignardAzur commented on masonry/src/contexts.rs:
```diff
@@ -879,6 +902,19 @@ impl LayoutCtx<'_> {
+    }
+
+    pub fn set_child_translation(&mut self, translation: Vec2) {
+        if self.widget_state.translation != translation {
```
Nah. Worst case, you do superfluous work.

PoignardAzur commented on masonry/src/passes/compose.rs:
```diff
@@ -0,0 +1,103 @@
+                panic!(
+                    "Error in '{}' #{}: cannot find child #{} returned by children_ids()",
+                    parent_name,
+                    parent_id.to_raw(),
```
Sure, but that's a future PR.

PoignardAzur commented on masonry/src/passes/compose.rs:
```diff
@@ -0,0 +1,103 @@
+) {
+    let moved = parent_moved || state.item.translation_changed;
+    let translation = parent_translation + state.item.translation + state.item.origin.to_vec2();
+    state.item.window_origin = Point::new(translation.x, translation.y);
```
(For what it's worth, you can create an inline suggestion with the "Add a suggestion button"; it's the one with the file icon with a +-, next to the "Preview" tab.)

PoignardAzur commented on masonry/src/contexts.rs:
```diff
@@ -6,6 +6,7 @@
 use std::time::Duration;
 
 use accesskit::{NodeBuilder, TreeUpdate};
+use kurbo::Vec2;
```
I think it doesn't make a difference, they resolve to the same symbol. (Otherwise I think rustc would complain.)

I don't know if we have a coding style for this, but either way, we're already using both version throughout Masonry, so I'll keep the ones you pointed out as-is.

PoignardAzur replied:
I think it doesn't make a difference, they resolve to the same symbol. (Otherwise I think rustc would complain.)

I don't know if we have a coding style for this, but either way, we're already using both version throughout Masonry, so I'll keep the ones you pointed out as-is.

DJMcNab commented on masonry/src/widget/widget_state.rs:
```diff
@@ -61,6 +67,11 @@ pub struct WidgetState {
     pub(crate) is_portal: bool,
 
+    pub(crate) request_compose: bool,
+    // TODO - Handle matrix transforms
```
Matrices are tricky with text, due to optical size of text (which we don't yet handle). I think changing the scale of a child can only correctly happen in `layout`

PoignardAzur replied:
I'm leaving that debate for once we do a complete layout refactor.

DJMcNab reviewed (commented):
It's hard to judge exactly what this is doing with no widgets actually using it...

Code seems about as expected, I believe.

PoignardAzur commented:
I think this is ready to merge?

Philipp-M commented on masonry/src/contexts.rs:
```diff
@@ -847,7 +870,7 @@ impl LayoutCtx<'_> {
         if origin != self.get_child_state_mut(child).origin {
             self.get_child_state_mut(child).origin = origin;
-            self.get_child_state_mut(child).needs_window_origin = true;
+            self.get_child_state_mut(child).translation_changed = true;
```
would `origin_changed` (or `window_origin_changed`) make more sense (as this is the defacto end-result if understand this correctly)?

DJMcNab replied:
AIUI, this flag means that "my local translation has changed", and so the "global translation" needs to be updated.

PoignardAzur replied:
Yeah, `translation` and `window_origin` have different semantics.

Philipp-M commented on masonry/src/contexts.rs:
```diff
@@ -859,6 +882,19 @@ impl LayoutCtx<'_> {
+        self.widget_state.needs_compose
+    }
+
+    pub fn set_child_translation(&mut self, translation: Vec2) {
```
```suggestion
    /// offsets the child by `translation` relative to the parents `window_origin`
    pub fn set_child_translation(&mut self, translation: Vec2) {
```

Philipp-M replied:
(Don't want this to be a conflict point, but was this overlooked or do you want some different doc-comment?)

PoignardAzur replied:
Huh, that's odd, I thought I'd added a comment there.

... And, looking again, yes I did. The UI is just absolutely terrible at showing it.

Philipp-M replied:
Ok, yeah indeed weird, I thought it would show `Outdated` or something, thanks :)

Philipp-M commented on masonry/src/passes/compose.rs:
```diff
@@ -0,0 +1,103 @@
+        },
+    );
+
+    // state.item.moved = false;
```
Nit: does this still make sense?

PoignardAzur replied:
Nope.

Philipp-M commented on masonry/src/tree_arena.rs:
```diff
@@ -372,6 +372,38 @@ impl<'a, Item> ArenaMutChildren<'a, Item> {
 }
 
+impl<'a, Item> ArenaRef<'a, Item> {
+    pub fn id(&self) -> u64 {
```
As this is `pub` I think it could have a `# Panics` doc-comment, that this could panic under certain conditions (same below for `ArenaMut`).

PoignardAzur replied:
The id is never None is that context. Added a comment to document it.

Philipp-M commented on masonry/src/widget/widget.rs:
```diff
@@ -105,6 +106,8 @@ pub trait Widget: AsAny {
     /// The layout strategy is strongly inspired by Flutter.
     fn layout(&mut self, ctx: &mut LayoutCtx, bc: &BoxConstraints) -> Size;
 
+    fn compose(&mut self, ctx: &mut ComposeCtx) {}
```
Since you add this already as default impl, should I make a PR for all the other methods here too? (After merging this, to avoid merge-conflicts)

PoignardAzur replied:
I'm adding the default implementations as I convert the passes. I'd rather not add them for passes that still need to recurse on children FWIW.

Philipp-M commented on masonry/src/widget/widget_pod.rs:
```diff
@@ -372,8 +372,9 @@ impl<W: Widget> WidgetPod<W> {
                     }
                 }
                 InternalLifeCycle::ParentWindowOrigin { .. } => {
```
As this doesn't do anything anymore, can we just remove this?

DJMcNab commented on masonry/src/tree_arena.rs:
```diff
@@ -372,6 +372,38 @@ impl<'a, Item> ArenaMutChildren<'a, Item> {
+        self.children.id.unwrap()
+    }
+
+    /// Returns a shared token equivalent to this one.
```
`Shared` doesn't seem quite right, given that `ArenaRef` cannot be copied (etc.). That is, `ArenaRef` does not appear to have any additional abilities over `ArenaMut`, except for being passed to methods which accept `ArenaRef`.

This method also doesn't seem to be used?

PoignardAzur replied:
> Shared doesn't seem quite right, given that ArenaRef cannot be copied

Well now it can =P

> This method also doesn't seem to be used?

Yeah, but I'm writing this as though it was a public crate.

DJMcNab commented on masonry/src/widget/widget_pod.rs:
```diff
@@ -372,8 +372,9 @@ impl<W: Widget> WidgetPod<W> {
                 InternalLifeCycle::ParentWindowOrigin { .. } => {
-                    state.parent_window_origin = parent_ctx.widget_state.window_origin();
-                    state.needs_window_origin = false;
+                    // FIXME
```
Does this event need to be removed?

DJMcNab reviewed (commented):
I think the old `ParentWindowOrigin` code can be removed now?
I would much prefer that to happen in this PR if so.

PoignardAzur commented on masonry/src/contexts.rs:
```diff
@@ -462,6 +473,17 @@ impl_context_method!(WidgetCtx<'_>, EventCtx<'_>, LifeCycleCtx<'_>, {
+    /// Request a compose pass.
+    ///
+    /// The compose pass is often cheaper than the layout pass, because it can only transform individual widgets' position.
+    /// [`compose`]: crate::Widget::compose
```
Fixed.

Philipp-M replied:
Hmm the Link seems still not to be working using `cargo doc --no-deps`

PoignardAzur commented on masonry/src/passes/compose.rs:
```diff
@@ -0,0 +1,103 @@
+// Copyright 2024 the Xilem Authors
+// SPDX-License-Identifier: Apache-2.0
+
+use kurbo::{Point, Vec2};
```
Won't fix, see other comment.

PoignardAzur commented:
I've done some small fixes to address code review. Next step is removing `Lifecycle::ParentWindowOrigin` and associated stuff.

PoignardAzur commented:
I believe this is ready to merge.

PoignardAzur commented:
We should probably merge #223 first, then rebase this on top of it.

PoignardAzur commented:
> We should probably merge https://github.com/linebender/xilem/pull/223 first, then rebase this on top of it.

Done. I believe this addresses all code review points.

DJMcNab reviewed (approved):
This is compelling! It's really nice to see the back of the parent window origin things

Philipp-M commented on masonry/src/contexts.rs:
```diff
@@ -412,6 +423,17 @@ impl_context_method!(MutateCtx<'_>, EventCtx<'_>, LifeCycleCtx<'_>, {
+    // TODO - Document better
+    /// Request a [`compose`] pass.
+    ///
+    /// The compose pass is often cheaper than the layout pass, because it can only transform individual widgets' position.
```
```suggestion
    /// The compose pass is often cheaper than the layout pass, because it can only transform individual widget's position.
```

(As I just see this)

PoignardAzur replied:
I think the old version is fine?

... you know what, let me look it up.

DJMcNab replied:
I think the original is the correct form for "the position of each of multiple widgets". However,
```suggestion
    /// The compose pass is often cheaper than the layout pass, because it can only change the position of widgets.
```
might be clearer still

PoignardAzur commented:
... It just now occurs to me, as I start the paint pass, that I completely forgot to hook composition with the Scene transforms. We didn't catch it because widgets don't use the pass yet.

Oh well, I'll hook it up with the Portal widget while I implement the paint pass.
